storage/src/mmr/bitmap.rs

```diff
@@ -405,41 +409,36 @@ impl<H: CHasher, const N: usize> Bitmap<H, N> {
         } else {
             chunk[byte_offset] &= !mask;
         }
-        if is_last {
-            // No need to insert into dirty_chunks since this bit is within the last (yet to be
-            // fully filled) chunk.
-            return;
+
+        // If the updated chunk is already in the MMR, mark it as dirty.
+        if chunk_index < self.authenticated_len {
+            self.dirty_chunks.insert(chunk_index);
         }
-        self.dirty_chunks.insert(chunk_index);
     }
 
-    /// Whether there are any unprocessed batch updates.
+    /// Whether there are any updates that are not yet reflected in this bitmap's root.
     pub fn is_dirty(&self) -> bool {
-        !self.dirty_chunks.is_empty()
+        !self.dirty_chunks.is_empty() || self.authenticated_len < self.bitmap.len() - 1
     }
 
-    /// Process all updates made since the last call to this method, allowing for `root` and proof
-    /// generation.
+    /// Process all updates not yet reflected in the bitmap's root.
     pub async fn sync(&mut self, hasher: &mut impl Hasher<H>) -> Result<(), Error> {
-        let mut chunks = self.dirty_chunks.iter().copied().collect::<Vec<_>>();
-        self.dirty_chunks.clear();
-        // We must process chunks in sorted order to ensure the underlying MMR is updated
-        // appropriately.
-        chunks.sort();
-
-        for chunk_index in chunks {
-            let leaf_pos = leaf_num_to_pos((chunk_index + self.pruned_chunks) as u64);
-            if leaf_pos < self.mmr.size() {
-                self.mmr
-                    .update_leaf_batched(hasher, leaf_pos, &self.bitmap[chunk_index])
-                    .await?;
-                continue;
-            }
-            assert_eq!(leaf_pos, self.mmr.size(), "invalid leaf pos");
-            self.mmr
-                .add_batched(hasher, &self.bitmap[chunk_index])
-                .await?;
+        // Add newly appended chunks to the MMR (other than the very last).
+        let start = self.authenticated_len;
+        let end = self.bitmap.len() - 1; // can't underflow -- always at least one chunk
```

Comment:
nit: seems like a good place to throw an assert rather than a comment

Reply:
done